### PR TITLE
docs: document rpc api versioning process.

### PR DESCRIPTION
**Pull request by dnldd**
This documents the process expected to be followed by contributors in versioning JSON-RPC API endpoints.

work towards #2746.

### CODE REVIEW

**Comment by matthawkins90** [comment]
I think this update is awesome. Had a quick question.

From the previous issue, the two main approaches are:

1. A single versioned endpoint for the entire RPC API, or
2. Individual versioned endpoints for each RPC method.

I just want to clarify what you mean by "Per-RPC Versioning" and "versioned per-RPC endpoint" When you say those words, are you referring to the 2nd approach above, or just that the RPC API will _have versions_ in general?

**Comment by dnldd** [comment]
After discussing the two approaches in chat with davec, versioning each RPC individually was the better approach and that is what this versioning process recommends. I will update the text to clearly indicate this where it unclear. I was supposed to comment on the issue regarding the preferred approach before writing the docs, missed a step there.

File: docs/rpc_api_versioning.md [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+ explicitly accessible via `existsaddressv1`.
+
+ - The default method name, `existsaddress` will resolve to the current version 
+ of the endpoint which is `existsaddressv2` in this case.
+
+ This scheme will be enforced via the JSON-RPC type and handler registrations. 
+
+<a name="Example" />
+
+### 4. Versioned Enpoint Example
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:63]: Typo: Endpoint

File: docs/rpc_api_versioning.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+	"existsaddress":         {(*bool)(nil)},
+	"existsaddressv1":       {(*bool)(nil)},
+	"existsaddressv2":       {(*bool)(nil)},
+    // ...
+}
+```
+
+<a name="DeprecationAndRemovalScheme" />
+
+### 5. Versioned Enpoint Deprecation And Removal Scheme
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:141]: Typo: Endpoint

File: docs/rpc_api_versioning.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+Only two versioned endpoints will be maintained per JSON-RPC endpoint. The
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:143]: "Each JSON-RPC method will only have two maintained endpoint versions. The introduction of a new endpoint version should immediately indicate that the oldest maintained endpoint will be deprecated. This is expected...."

File: docs/rpc_api_versioning.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+3. [Versioned Endpoint Accessibilty Scheme](#AccessibiltyScheme)<br />
+4. [Versioned Enpoint Example](#Example)<br />
+5. [Versioned Enpoint Deprecation And Removal Scheme](#DeprecationAndRemovalScheme)<br />
+6. [Versioned Endpoint JSON-RPC Unit Testing](#UnitTesting)<br />
+7. [Versioned JSON-RPC Documentation](#Documentation)<br />
+
+
+<a name="PerRPCVersioning" />
+
+### 1. Per-RPC Versioning
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:13]: `### 1. Per-method Versioning`
(see below for reasoning)

File: docs/rpc_api_versioning.md [hunk 5]
```diff
@@ -0,0 +1,272 @@
+### Table of Contents
+1. [Per-RPC Versioning](#PerRPCVersioning)<br />
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:2]: Per-method Versioning......#PerMethodVersioning
(see below for reasoning)

File: docs/rpc_api_versioning.md [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+pull request introducing the new versioned endpoint. Using `existsaddress` as 
+an example, with the introduction of the versioned endpoint `existsaddressv2` 
+`existsaddressv1` will be marked as deprecated and eventually removed when 
+the next versioned endpoint `existsaddressv3` is introduced.
+
+<a name="UnitTesting" />
+
+### 6. Versioned Endpoint JSON-RPC Unit Testing
+
+All versioned JSON-RPC handlers are required to have their own set of unit tests.
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:155]: ~...versioned JSON-RPC handlers...~
...versioned method handlers...

File: docs/rpc_api_versioning.md [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    // ...
+    })
+}
+
+// ...
+```
+
+<a name="Documentation" />
+
+### 7. Versioned JSON-RPC Documentation
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:201]: `### 7. Versioned Endpoint Documentation`

File: docs/rpc_api_versioning.md [hunk 8]
```diff
@@ -0,0 +1,272 @@
+### Table of Contents
+1. [Per-RPC Versioning](#PerRPCVersioning)<br />
+2. [Versioning Scheme](#VersioningScheme)<br />
+3. [Versioned Endpoint Accessibilty Scheme](#AccessibiltyScheme)<br />
+4. [Versioned Enpoint Example](#Example)<br />
+5. [Versioned Enpoint Deprecation And Removal Scheme](#DeprecationAndRemovalScheme)<br />
+6. [Versioned Endpoint JSON-RPC Unit Testing](#UnitTesting)<br />
+7. [Versioned JSON-RPC Documentation](#Documentation)<br />
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:8]: `7. [Versioned Endpoint Documentation](#Documentation)<br />`

File: docs/rpc_api_versioning.md [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+// ...
+```
+
+<a name="Documentation" />
+
+### 7. Versioned JSON-RPC Documentation
+
+Documentation for the newly versioned JSON-RPC endpoint will be expected in the
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:203]: ~JSON-RPC~

**Review by matthawkins90 — COMMENTED** [review]
Looking really good, and I like this schema.
Hopefully my comments are clear and helpful!

File: docs/rpc_api_versioning.md [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+5. [Versioned Enpoint Deprecation And Removal Scheme](#DeprecationAndRemovalScheme)<br />
+6. [Versioned Endpoint JSON-RPC Unit Testing](#UnitTesting)<br />
+7. [Versioned JSON-RPC Documentation](#Documentation)<br />
+
+
+<a name="PerRPCVersioning" />
+
+### 1. Per-RPC Versioning
+
+To streamline API access in the process of transitioning between release
```
[the remark this answers is not in the document]
Comment (dnldd) [reply docs/rpc_api_versioning.md:15]: Yeah, I think that's clearer. Will make the changes.

**Comment by dnldd** [comment]
I'm going to leave this in draft still and look to add a section for versioning websocket endpoints.

File: docs/rpc_api_versioning.md [hunk 11]
```diff
@@ -269,6 +270,88 @@ the JSON-RPC API documentation will be updated to include documentation for
     // ...
 ```
 
+<a name="Websockets" />
+
+### 8. Websocket Endpoints & Example
+Websocket endpoints will be versioned using the accessibilty scheme outlined
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:276]: Typo: accessibility

File: docs/rpc_api_versioning.md [hunk 12]
```diff
@@ -269,6 +270,88 @@ the JSON-RPC API documentation will be updated to include documentation for
...
+func handleRescanV2(wsc *wsClient, icmd interface{}) (interface{}, error) {
+	cmd, ok := icmd.(*types.RescanV2Cmd)
+	if !ok {
+		return nil, dcrjson.ErrRPCInternal
+	}
+    // ...
+}
+```
+
 With the outlined schemes adhered to, the versioned endpoint should be ready
```
Comment (matthawkins90) [inline docs/rpc_api_versioning.md:355]: This last sentence is clunky, and I'm not sure what you're trying to say. Is this a list of remaining steps to take before the new versioned endpoints API can be active? I don't really understand the second half of the sentence, but I gave it my best shot. Please feel free to modify:

Once all `dcrd` JSON-RPC API methods are updated using the schema outlined above, the versioned endpoints will be ready for use. However, `dcrctl` and other consumers will need to update their `rpc/jsonrpc/types` dependencies and logic.

**Review by matthawkins90 — COMMENTED** [review]
I think the comments and structure makes sense. Unfortunately, I'm only a novice with golang, so I can't really comment on the code examples.

File: docs/rpc_api_versioning.md (hunk 12 above)
Comment (dnldd) [reply docs/rpc_api_versioning.md:355]: I'm saying dcrctl needs to be rebuilt with the updated jsonrpc types to be able to access the new versioned commands.

**Comment by dnldd** [comment]
Just realized having the default method name resolve to the current version of the endpoint will break consumers once they update. We don't want that, should be resolving to the older version instead. Will make the needed changes soon.

**Review by matthawkins90 — APPROVED** [review]

**Pull request closed without merging**